lib/rust/lexer/definition/src/lexer.rs

```diff
@@ -177,6 +189,192 @@ impl EnsoLexer {
+            let remaining_indent = " ".repeat(indent - comment_indent);
+            self.comment_state.append_to_line(remaining_indent);
+        }
+    }
+
+    /// Triggered when an empty line is discovered in a doc comment.
+    fn doc_comment_on_empty_line<R:LazyReader>
+    ( &mut self
+    , line_ending:token::LineEnding
+    , reader:&mut R
+    ) {
+        trace!(self.logger,"Comment::doc_comment_on_empty_line");
+        let current = self.consume_current();
+        let indent  = current.chars().count() - line_ending.size();
+        self.comment_state.submit_blank_line(indent,line_ending);
+        if let token::LineEnding::None = line_ending {
+            let block_newline = self.block_newline;
+            self.on_end_doc_comment(reader);
+            self.push_state(block_newline);
+            self.on_eof(reader);
+        }
+    }
+
+    // === Comment Rules ===
+
+    /// The rules for lexing Enso comments.
+    fn add_comment_rules(lexer:&mut EnsoLexer) {
+        let comment_char      = c!('#');
+        let lf                = l!("\n");
+        let crlf              = l!("\r\n");
+        let eof               = Pattern::eof();
+        let spaces            = EnsoLexer::spaces();
+        let doc_comment_start = &comment_char >> &comment_char >> &spaces;
+        let eof_newline       = &spaces.opt() >> &eof;
+        let empty_line_lf     = &spaces.opt() >> &lf;
+        let empty_line_crlf   = &spaces.opt() >> &crlf;
+
+        let initial_state_id = lexer.initial_state;
+        let initial_state    = lexer.group_mut(initial_state_id);
+        initial_state.create_rule(&doc_comment_start,"self.on_begin_doc_comment(reader)");
+        initial_state.create_rule(&comment_char,     "self.on_begin_disable_comment(reader)");
+
+        let disable_comment_id = lexer.disable_comment;
+        let disable_comment    = lexer.group_mut(disable_comment_id);
+        disable_comment.create_rule(&lf,"self.on_end_disable_comment(token::LineEnding::LF,reader)");
+        disable_comment.create_rule(&crlf,"self.on_end_disable_comment(token::LineEnding::CRLF,reader)");
+        disable_comment.create_rule(&eof,"self.on_end_disable_comment(token::LineEnding::None,reader)");
+        disable_comment.create_rule(&Pattern::any(),"self.on_build_disable_comment(reader)");
+
+        let doc_comment_id = lexer.doc_comment;
+        let doc_comment    = lexer.group_mut(doc_comment_id);
+        doc_comment.create_rule(&lf,"self.on_doc_comment_end_of_line(token::LineEnding::LF,reader)");
+        doc_comment.create_rule(&crlf,"self.on_doc_comment_end_of_line(token::LineEnding::CRLF,reader)");
+        doc_comment.create_rule(&eof,"self.on_doc_comment_end_of_line(token::LineEnding::None,reader)");
+        doc_comment.create_rule(&Pattern::any(),"self.on_build_doc_comment_line(reader)");
+
+        let doc_comment_newline_id = lexer.doc_comment_newline;
+        let doc_comment_newline    = lexer.group_mut(doc_comment_newline_id);
```

Comment:
we have a rule in GUI for a long time now - if yo uare making such newlines in a function, it means it is either too long and you need to divide it to several functions, or just add here sections in each division point to describe what is there

Reply:
Surely the fact that each section begins with a new state is documentation enough?

Reply:
Nope, its hard to read this way. Thsi is just to make it more clear. ITsj ust like keeping 100 chars long lines. Either divide it to well-named sub-functions, or add sections there. TBH when reading the code right now I have to spend much more time to understand parts of the code than just borwisgn sections.